From 27a929af1f71a2f5d9926fee84ac2f2c5e8e98e8 Mon Sep 17 00:00:00 2001
From: xGood <zenia2003725@gmail.com>
Date: Sun, 23 Mar 2025 15:46:18 +0300
Subject: [PATCH 3/4] multiplayer

---
 20250317/1/client.py |  13 ++++++
 20250317/1/prog.py   | 109 ++++++++++++-------------------------------
 2 files changed, 43 insertions(+), 79 deletions(-)

diff --git a/20250317/1/client.py b/20250317/1/client.py
index 1f0f332..522cecd 100644
--- a/20250317/1/client.py
+++ b/20250317/1/client.py
@@ -3,6 +3,7 @@ import sys
 import socket
 import shlex
 import json
+import cowsay
 
 
 WEAPONS_LIST = {"sword": 10, "spear": 15, "axe": 20}
@@ -44,6 +45,17 @@ class cmd_line(cmd.Cmd):
             x, y = int(rules[coords_ind+1]), int(rules[coords_ind+2])
             hp = int(rules[hp_ind+1])
             hello = rules[hello_ind+1]
+            try:
+                x, y = int(x), int(y)
+            except:
+                print("Invalid arguments")
+                return
+            if name not in [*cowsay.list_cows(), "jgsbat"]:
+                print("Cannot add unknown monster")
+                return
+            if x < 0 or y < 0 or x >= field_x or y >= field_y or not (hasattr(hello, "__str__") or hasattr(hello, "__repr__")):
+                print("Invalid arguments")
+                return
         except:
             print("Invalid arguments")
         self.socket.sendall(f"addmon {x} {y} {hp} {name} {hello}")
@@ -87,4 +99,5 @@ if __name__ == "__main__":
     port = 1337 if len(sys.argv) < 3 else int(sys.argv[2])
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
         s.connect((host, port))
+        field_x, field_y = list(map(int, s.recv(1024).decode().split()))
         cmd_line(sock=s).cmdloop()
diff --git a/20250317/1/prog.py b/20250317/1/prog.py
index 8f852f4..1869380 100644
--- a/20250317/1/prog.py
+++ b/20250317/1/prog.py
@@ -2,71 +2,15 @@ import cowsay
 from io import StringIO
 import shlex
 import cmd
+import socket
+import sys
+import json
 
 
 WEAPONS_LIST = {"sword": 10, "spear": 15, "axe": 20}
 dflt_wpn = "sword"
 
 
-class cmd_line(cmd.Cmd):
-
-    prompt = "(MUD) "
-
-    def do_up(self, args):
-        plr.move("up")
-    
-    def do_down(self, args):
-        plr.move("down")
-    
-    def do_left(self, args):
-        plr.move("left")
-    
-    def do_right(self, args):
-        plr.move("right")
-    
-    def do_addmon(self, args):
-        name, *rules = shlex.split(args)
-        try:
-            hello_ind = rules.index("hello")
-            hp_ind = rules.index("hp")
-            coords_ind = rules.index("coords")
-            x, y = int(rules[coords_ind+1]), int(rules[coords_ind+2])
-            hp = int(rules[hp_ind+1])
-            hello = rules[hello_ind]
-        except:
-            print("Invalid arguments")
-        fld.addmon(x, y, hp, name, hello)
-    
-    def do_attack(self, args):
-        try:
-            name = shlex.split(args)[0]
-        except:
-            print("Invalid command")
-            return
-        if "with" in args:
-            try:
-                name, _, weapon = shlex.split(args)
-                if WEAPONS_LIST.get(weapon, None) is None:
-                    print("Unknown weapon")
-                    return
-                plr.attack(name, WEAPONS_LIST[weapon])
-            except ValueError:
-                print("Need to specify the weapon")
-                return
-        else:
-            plr.attack(name, WEAPONS_LIST[dflt_wpn])
-    
-    def complete_attack(self, text, line, ind1, ind2):
-        words = shlex.split(line)
-        if len(words) == 2:
-            return [c for c in fld.monsters_dict.keys() if fld.monsters_dict[c] > 0 and c.startswith(text)]
-        if len(words) < 3:
-            return []
-        if len(words) == 3:
-            return WEAPONS_LIST.keys()
-        return [c for c in WEAPONS_LIST.keys() if c.startswith(text)]
-
-
 jgsbat = cowsay.read_dot_cow(StringIO("""
     ,_                    _,
     ) '-._  ,_    _,  _.-' (
@@ -97,18 +41,8 @@ class Field:
         return self._y
     
     def addmon(self, x, y, hp, name, msg):
-        try:
-            x, y = int(x), int(y)
-        except:
-            print("Invalid arguments")
-            return
-        if x < 0 or y < 0 or x >= self.x or y >= self.y or not (hasattr(msg, "__str__") or hasattr(msg, "__repr__")):
-            print("Invalid arguments")
-            return
-        if name not in [*cowsay.list_cows(), "jgsbat"]:
-            print("Cannot add unknown monster")
-            return
         self.monsters_dict[name] = self.monsters_dict.get(name, 0) + 1
+        conn.send(json.dumps(self.monsters_dict))
         self.field[x][y] = Monster(x, y, hp, name, msg)
 
 
@@ -122,9 +56,10 @@ class Player:
 
     def move(self, direction):
         self._x, self._y = (self._x + self.__class__.direct_map[direction][0]) % self.fld.x, (self._y + self.__class__.direct_map[direction][1]) % self.fld.y
-        print(f"Moved to ({self._x}, {self._y})")
-        if self.fld.field[self._x][self._y]:
-            encounter(self._x, self._y, self.fld.field)
+       conn.send(f"Moved to ({self._x}, {self._y})\n".encode())
+       if self.fld.field[self._x][self._y]:
+            conn.send(f"Found {self.fls.field[self._x][self._y].name} {self.fls.field[self._x][self._y]._msg}\n".encode())
+
 
     def attack(self, name, damage):
         if self.fld.field[self._x][self._y] and self.fld.field[self._x][self._y].name == name:
@@ -132,7 +67,7 @@ class Player:
             if result:
                 del self.fld.field[self._x][self._y]
             return
-        print(f"No {name} here")
+        conn.send(f"No {name} here\n".encode())
         return
 
 
@@ -145,7 +80,7 @@ class Monster:
     def __init__(self, x, y, hp, name, msg, func=None):
         self._x, self._y, self.name, self._msg, self._func = x, y, name, msg, func
         self._hp = hp
-        print(f"Added monster {name} to ({x}, {y}) saying {msg}")
+        conn.send(f"Added monster {name} to ({x}, {y}) saying {msg}\n".encode())
         if self._func is None:
             if name == "jgsbat":
                 self._func = lambda x : print(cowsay.cowsay(x, cowfile=jgsbat))
@@ -159,7 +94,7 @@ class Monster:
         return True
 
     def attacked(self, damage):
-        print(f"Attacked {self.name}, damage {min(damage, self._hp)}")
+        conn.send(f"Attacked {self.name}, damage {min(damage, self._hp)}\n")
         self._hp -= min(damage, self._hp)
         if self._hp == 0:
             print(f"{self.name} died")
@@ -169,9 +104,25 @@ class Monster:
             return False
 
 
-
 if __name__ == "__main__":
     fld = Field(10, 10)
     plr = Player(fld)
-    print("<<< Welcome to Python-MUD 0.1 >>>")
-    cmd_line().cmdloop()
+    host = "localhost" if len(sys.argv) < 2 else sys.argv[1]
+    port = 1337 if len(sys.argv) < 3 else int(sys.argv[2])
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind((host, port))
+        s.listen()
+        conn, addr = s.accept()
+        with conn:
+            conn.sendall(f"{10} {10}".encode())
+            print('Connected by', addr)
+            while data := conn.recv(1024):
+                info = shlex.split(data.decode())
+                print(info)
+                if info[0] == "move":
+                    plr.move(info[1])
+                if info == ["info", "host"]:
+                    print(addr[0])
+                    conn.sendall(addr[0].encode())
+                if info == ["info", "port"]:
+                    conn.sendall(str(addr[1]).encode())
-- 
2.34.1

